docs/ci_parse_ptr_reply.3
.\"
.\" Copyright 1998 by the Massachusetts Institute of Technology.
.\" SPDX-License-Identifier: MIT
.\"
.TH CI_PARSE_PTR_REPLY 3 "25 July 1998"
.SH NAME
ci_parse_ptr_reply \- Parse a reply to a DNS query of type PTR into a hostent
.SH SYNOPSIS
.nf
#include <ci.h>

int ci_parse_ptr_reply(const unsigned char *\fIabuf\fP, int \fIalen\fP,
                         const void *\fIaddr\fP, int \fIaddrlen\fP,
                         int \fIfamily\fP, struct hostent **\fIhost\fP);
.fi
.SH DESCRIPTION
The
.B ci_parse_ptr_reply
function parses the response to a query of type PTR into a
.BR "struct hostent" .
The parameters
.I abuf
and
.I alen
give the contents of the response.  The parameters
.IR addr ,
.IR addrlen ,
and
.I family
specify which address was queried for; they are not used to verify the
response, merely used to fill in the address of the
.BR "struct hostent" .
The resulting
.B struct hostent
is stored in allocated memory and a pointer to it stored into the
variable pointed to by
.IR host .
It is the caller's responsibility to free the resulting host structure
using
.BR ci_free_hostent (3)
when it is no longer needed.
.SH RETURN VALUES
.B ci_parse_ptr_reply
can return any of the following values:
.TP 15
.B CI_SUCCESS
The response was successfully parsed.
.TP 15
.B CI_EBADRESP
The response was malformatted.
.TP 15
.B CI_ENODATA
The response did not contain an answer to the query.
.TP 15
.B CI_ENOMEM
Memory was exhausted.
.SH SEE ALSO
.BR ci_gethostbyaddr (3),
.BR ci_free_hostent (3)


RELEASE-NOTES.md
## c-ci version 1.34.2 - October 15 2024

This release contains a fix for downstream packages detecting the c-ci
version based on the contents of the header file rather than the
distributed pkgconf or cmake files.

## c-ci version 1.34.1 - October 9 2024

This release fixes a packaging issue.


## c-ci version 1.34.0 - October 9 2024

This is a feature and bugfix release.

Features:
* adig: read arguments from adigrc.
  [PR #856](https://github.com/c-ci/c-ci/pull/856)
* Add new pending write callback optimization via `ci_set_pending_write_cb`.
  [PR #857](https://github.com/c-ci/c-ci/pull/857)
* New function `ci_process_fds()`.
  [PR #875](https://github.com/c-ci/c-ci/pull/875)
* Failed servers should be probed rather than redirecting queries which could
  cause unexpected latency.
  [PR #877](https://github.com/c-ci/c-ci/pull/877)
* adig: rework command line arguments to mimic dig from bind.
  [PR #890](https://github.com/c-ci/c-ci/pull/890)
* Add new method for overriding network functions
  `ci_set_socket_function_ex()` to properly support all new functionality.
  [PR #894](https://github.com/c-ci/c-ci/pull/894)
* Fix regression with custom socket callbacks due to DNS cookie support.
  [PR #895](https://github.com/c-ci/c-ci/pull/895)
* ci_socket: set IP_BIND_ADDRESS_NO_PORT on ci_set_local_ip* tcp sockets
  [PR #887](https://github.com/c-ci/c-ci/pull/887)
* URI parser/writer for ci_set_servers_csv()/ci_get_servers_csv().
  [PR #882](https://github.com/c-ci/c-ci/pull/882)

Changes:
* Connection handling modularization.
  [PR #857](https://github.com/c-ci/c-ci/pull/857),
  [PR #876](https://github.com/c-ci/c-ci/pull/876)
* Expose library/utility functions to tools.
  [PR #860](https://github.com/c-ci/c-ci/pull/860)
* Remove `ci__` prefix, just use `ci_` for internal functions.
  [PR #872](https://github.com/c-ci/c-ci/pull/872)


Bugfixes:
* fix: potential WIN32_LEAN_AND_MEAN redefinition.
  [PR #869](https://github.com/c-ci/c-ci/pull/869)
* Fix googletest v1.15 compatibility.
  [PR #874](https://github.com/c-ci/c-ci/pull/874)
* Fix pkgconfig thread dependencies.
  [PR #884](https://github.com/c-ci/c-ci/pull/884)


Thanks go to these friendly people for their efforts and contributions for this
release:

* Brad House (@bradh352)
* Cristian Rodríguez (@crrodriguez)
* Georg (@tacerus)
* @lifenjoiner
* Shelley Vohr (@codebytere)
* 前进，前进，进 (@leleliu008)



src/lib/legacy/ci_parse_ptr_reply.c
/* MIT License
 *
 * Copyright (c) 2023 Brad House
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice (including the next
 * paragraph) shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * SPDX-License-Identifier: MIT
 */

#include "ci_private.h"

#ifdef HAVE_NETINET_IN_H
#  include <netinet/in.h>
#endif
#ifdef HAVE_NETDB_H
#  include <netdb.h>
#endif


ci_status_t ci_parse_ptr_reply_dnsrec(const ci_dns_record_t *dnsrec,
                                          const void *addr, int addrlen,
                                          int family, struct hostent **host)
{
  ci_status_t   status;
  size_t          ptrcount = 0;
  struct hostent *hostent  = NULL;
  const char     *hostname = NULL;
  const char     *ptrname  = NULL;
  size_t          i;
  size_t          ancount;

  *host = NULL;

  /* Fetch name from query as we will use it to compare later on.  Old code
   * did this check, so we'll retain it. */
  status = ci_dns_record_query_get(dnsrec, 0, &ptrname, NULL, NULL);
  if (status != CI_SUCCESS) {
    goto done;
  }

  ancount = ci_dns_record_rr_cnt(dnsrec, CI_SECTION_ANSWER);
  if (ancount == 0) {
    status = CI_ENODATA;
    goto done;
  }

  /* Response structure */
  hostent = ci_malloc(sizeof(*hostent));
  if (hostent == NULL) {
    status = CI_ENOMEM;
    goto done;
  }

  memset(hostent, 0, sizeof(*hostent));

  hostent->h_addr_list = ci_malloc(2 * sizeof(*hostent->h_addr_list));
  if (hostent->h_addr_list == NULL) {
    status = CI_ENOMEM;
    goto done;
  }
  memset(hostent->h_addr_list, 0, 2 * sizeof(*hostent->h_addr_list));
  if (addr != NULL && addrlen > 0) {
    hostent->h_addr_list[0] = ci_malloc((size_t)addrlen);
    if (hostent->h_addr_list[0] == NULL) {
      status = CI_ENOMEM;
      goto done;
    }
    memcpy(hostent->h_addr_list[0], addr, (size_t)addrlen);
  }
  hostent->h_addrtype = (HOSTENT_ADDRTYPE_TYPE)family;
  hostent->h_length   = (HOSTENT_LENGTH_TYPE)addrlen;

  /* Preallocate the maximum number + 1 */
  hostent->h_aliases = ci_malloc((ancount + 1) * sizeof(*hostent->h_aliases));
  if (hostent->h_aliases == NULL) {
    status = CI_ENOMEM;
    goto done;
  }
  memset(hostent->h_aliases, 0, (ancount + 1) * sizeof(*hostent->h_aliases));


  /* Cycle through answers */
  for (i = 0; i < ancount; i++) {
    const ci_dns_rr_t *rr =
      ci_dns_record_rr_get_const(dnsrec, CI_SECTION_ANSWER, i);

    if (rr == NULL) {
      /* Shouldn't be possible */
      status = CI_EBADRESP; /* LCOV_EXCL_LINE: DefensiveCoding */
      goto done;              /* LCOV_EXCL_LINE: DefensiveCoding */
    }

    if (ci_dns_rr_get_class(rr) != CI_CLASS_IN) {
      continue;
    }

    /* Any time we see a CNAME, replace our ptrname with its value */
    if (ci_dns_rr_get_type(rr) == CI_REC_TYPE_CNAME) {
      ptrname = ci_dns_rr_get_str(rr, CI_RR_CNAME_CNAME);
      if (ptrname == NULL) {
        status = CI_EBADRESP; /* LCOV_EXCL_LINE: DefensiveCoding */
        goto done;              /* LCOV_EXCL_LINE: DefensiveCoding */
      }
    }

    /* Handling for PTR records below this, otherwise skip */
    if (ci_dns_rr_get_type(rr) != CI_REC_TYPE_PTR) {
      continue;
    }

    /* Issue #683
     * Old code compared the name in the rr to the ptrname, but I think this
     * is wrong since it was proven wrong for A & AAAA records.  Leaving
     * this code commented out for future reference
     *
     * rname = ci_dns_rr_get_name(rr);
     * if (rname == NULL) {
     *   status = CI_EBADRESP;
     *   goto done;
     * }
     * if (!ci_strcaseeq(ptrname, rname)) {
     *   continue;
     * }
     */

    /* Save most recent PTR record as the hostname */
    hostname = ci_dns_rr_get_str(rr, CI_RR_PTR_DNAME);
    if (hostname == NULL) {
      status = CI_EBADRESP; /* LCOV_EXCL_LINE: DefensiveCoding */
      goto done;              /* LCOV_EXCL_LINE: DefensiveCoding */
    }

    /* Append as an alias */
    hostent->h_aliases[ptrcount] = ci_strdup(hostname);
    if (hostent->h_aliases[ptrcount] == NULL) {
      status = CI_ENOMEM;
      goto done;
    }
    ptrcount++;
  }

  if (ptrcount == 0) {
    status = CI_ENODATA;
    goto done;
  } else {
    status = CI_SUCCESS;
  }

  /* Fill in hostname */
  hostent->h_name = ci_strdup(hostname);
  if (hostent->h_name == NULL) {
    status = CI_ENOMEM; /* LCOV_EXCL_LINE: OutOfMemory */
    goto done;            /* LCOV_EXCL_LINE: OutOfMemory */
  }

done:
  if (status != CI_SUCCESS) {
    ci_free_hostent(hostent);
    /* Compatibility */
    if (status == CI_EBADNAME) {
      status = CI_EBADRESP;
    }
  } else {
    *host = hostent;
  }
  return status;
}

int ci_parse_ptr_reply(const unsigned char *abuf, int alen_int,
                         const void *addr, int addrlen, int family,
                         struct hostent **host)
{
  size_t             alen;
  ci_dns_record_t *dnsrec = NULL;
  ci_status_t      status;

  if (alen_int < 0) {
    return CI_EBADRESP;
  }

  alen = (size_t)alen_int;

  status = ci_dns_parse(abuf, alen, 0, &dnsrec);
  if (status != CI_SUCCESS) {
    goto done;
  }

  status = ci_parse_ptr_reply_dnsrec(dnsrec, addr, addrlen, family, host);

done:
  ci_dns_record_destroy(dnsrec);
  if (status == CI_EBADNAME) {
    status = CI_EBADRESP;
  }
  return (int)status;
}
